### DISCUSSION TITLE
How can I get controls of given class name inside an invisible templatedcontrol?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#14124] [Q&A]
What I want to achieve is to get all TextBlocks in MainWindow that include the specified class.

I am using the way introduced  here:https://github.com/AvaloniaUI/Avalonia/issues/3085

Here is the Templated Control:
```
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:controls="using:AvaloniaApplication3">
  <Design.PreviewWith>
    <controls:TestThumb />
  </Design.PreviewWith>

  <Style Selector="controls|TestThumb">
    <!-- Set Defaults -->
    <Setter Property="Template">
      <ControlTemplate>
		  <Border Background="{TemplateBinding Background}">
			  <ContentPresenter Margin="{TemplateBinding Padding}" Content="{TemplateBinding Content}">
			  </ContentPresenter>
		  </Border>
      </ControlTemplate>
    </Setter>
  </Style>
</Styles>

```

```
using Avalonia;
using Avalonia.Controls.Primitives;
using Avalonia.Metadata;

namespace AvaloniaApplication3;

public class TestThumb : Thumb
{
    public static readonly StyledProperty<object> ContentProperty =
 AvaloniaProperty.Register<TestThumb, object>(nameof(Content));
    [Content]
    public object Content
    {
        get
        {
            return GetValue(ContentProperty);
        }
        set
        {
            SetValue(ContentProperty, value);
        }
    }
}
```

Here is the MainWindow:
```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
		xmlns:Local="using:AvaloniaApplication3"
        x:Class="AvaloniaApplication3.MainWindow"
        Title="AvaloniaApplication3">
	<Local:TestThumb Width="200" Height="200" Background="Red" IsVisible="False">
		<TextBlock>123</TextBlock>
	</Local:TestThumb>
</Window>
```
```
using Avalonia.Controls;
using Avalonia.Media;
using Avalonia.VisualTree;
using System.Linq;

namespace AvaloniaApplication3
{
    public partial class MainWindow : Window
    {
        public MainWindow()
        {
            InitializeComponent();
            this.Loaded += MainWindow_Loaded;
        }

        private void MainWindow_Loaded(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
        {
            var i = this.GetVisualDescendants().OfType<TextBlock>().ToList();
        }
    }
}
```

If I set the IsVisible of Templated Control to false, the result will be like this:
![image](https://github.com/AvaloniaUI/Avalonia/assets/30720036/d1b27c61-9a59-4522-abe8-831e31ccf71d)

However, if I set the IsVisible of Tempalted Control to true, the result will be like this:
![image](https://github.com/AvaloniaUI/Avalonia/assets/30720036/b2814b1b-f6e2-4fa7-8af0-4ec03c0614cf)

What's wrong with it? How can I achieve what I want (regardless of whether IsVisible is true or false)

**Comment by timunie** [comment]
not visible is not in the visual tree I guess. Check DevTools [F12] for it.

**Reply by kyuranger** [reply]
Yes, as you see, it doesn't exist in visual tree while IsVisible set false.
![image](https://github.com/AvaloniaUI/Avalonia/assets/30720036/bf69d68d-6d3d-4d1f-97c2-993ba2b8e96f)

And it exists in visual tree while IsVislble set true.
![image](https://github.com/AvaloniaUI/Avalonia/assets/30720036/a983eadf-6a7d-4406-a8ce-cdb8a3f0cafa)

How can I achieve what I need in this case? I am afraid I must set the default value of IsVisible to false in my case.

**Reply by kyuranger** [reply]
In fact, I am making something similar to an iPad menu. There are a bunch of controls in the menu (marked with arrows in the picture below) that need to be controlled by code to play animations. However, due to the irregular layout of these controls, I cannot use StackPanel or the like to host them. I can only achieve such a layout through different grids.

As mentioned above, I need to get these controls and play animations, so my idea is to mark each control with the same class, then get the collection of these controls at once through code, and then play the animation at the back-end.

These controls will be invisible while the menu doesn't drag down. Because of this, I have to set the default value of IsVisible to false.

![image](https://github.com/AvaloniaUI/Avalonia/assets/30720036/a63b4ff5-bee6-4670-a2c4-2b8db4af5c32)
